hypervisor/arch/x86/reloc.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2018 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <hypervisor.h>
+#include <reloc.h>
+
+struct Elf64_Dyn {
+	uint64_t d_tag;
+	uint64_t d_ptr;
+};
+
+#define DT_NULL		0	/* end of .dynamic section */
+#define DT_RELA		7	/* relocation table */
+#define DT_RELASZ	8	/* size of reloc table */
+#define DT_RELAENT	9	/* size of one entry */
+
+struct Elf64_Rel {
+	uint64_t r_offset;
+	uint64_t r_info;
+	uint64_t reserved;
+};
+
+#define ELF64_R_TYPE(i)		((i) & 0xffffffff)
+#define R_X86_64_RELATIVE	8
+
+/* get the delta between CONFIG_RAM_START and the actual load address */
+uint64_t get_hv_image_delta(void)
+{
+	uint64_t addr;
+
+	asm volatile (" call 0f\n"
+		"0: pop %%rax\n"
+		"	sub $0b, %%rax\n"
+		"	mov %%rax, %0\n"
+		: "=m" (addr)
+		:
+		: "%rax");
+
+	return addr;
+}
+
+/* get the actual Hypervisor load address */
+uint64_t get_hv_image_base(void)
+{
+	return (get_hv_image_delta() + CONFIG_RAM_START);
+}
+
+/*
+ * Because trampoline code is relocated in different way, if HV code
+ * accesses trampoline using relative addressing, it needs to take
+ * out the HV relocation delta
+ *
+ * This function is valid if:
+ *  - The hpa of HV code is always higher than trampoline code
+ *  - The HV code is always relocated to higher address, compared
+ *    with CONFIG_RAM_START
+ */
+uint64_t trampoline_sym(void *addr)
```

Comment:
why it is _sym? are u trying to get the address after relocation? trampoline_reloated_addr or reloated_trampoline_addr?

Reply:
Thanks for pointing it out. I'll change it to trampoline_reloc_addr